### PR TITLE
add deduplication of types

### PR DESCRIPTION
**Pull request by walter9388**
Relating to https://github.com/asottile/pyupgrade/issues/982

Took me a while to get round to this, but here we are....

I think there are three different levels of approaching this issue:
1. Remove extra `None` types only:
    ```diff
    -def f(x: Optional[Union[int, None]]): pass
    +def f(x: int | None): pass
    -def g(x: Union[Optional[int], None]): pass
    +def g(x: int | None): pass
    ```
2. Remove any duplicated scalar types at the same depth by name in `Union` blocks:
    ```diff
    -def f(x: Union[Union[Union[Union[a, b], c], d], a]): pass
    +def f(x: a | b | c | d): pass
    -def g(x: Union[a.b | a.c, a.b, list[str], str]): pass
    +def g(x: a.b | a.c | list[str] | str): pass
    ```
3. General deduplication at any depth on any block:
    ```diff
    -def f(x: Union[list[Union[int, str]], list[Union[str, int]]]): pass
    +def f(x: list[int | str]): pass
    ```

I settled on level 2 as this was still possible with a single pass and seemed more useful than just focusing on `None` types.
I couldn't see a way of approaching the general problem (level 3) without recursively making a tree structure and then assessing the leaf nodes. However, I am not deeply familiar with the standard python libraries for parsing ASTs etc., so if there are simple built in methods for problems like this I would be interested to know!

I used the existing scan in `_fix_union` to determine the delimitators at `depth==1` between the types. This seemed to work well, but I definitely ran into some interesting edge cases when it came to handling comments, whitespace and multilines.

I have managed to get this working for a variety of test cases, and I would be interested to hear your feedback.

Btw I enjoy your YouTube content! I have learnt a lot of niche things I would have struggled to pick up otherwise. So thank you for that!

### CODE REVIEW

File: tests/features/typing_pep604_test.py [hunk 1]
```diff
@@ -242,6 +242,84 @@ def f(x: int | str) -> None: ...
...
+            ')\n',
+
+            id='duplicated types in multi-line nested unions or optionals',
+        ),
+        pytest.param(
+            'from typing import Union\n'
+            'def f(x: Union[list[int, str], list[Union[str, int], int]]): pass\n',
+
+            'from typing import Union\n'
+            'def f(x: list[int, str]): pass\n',
```
Comment (asottile) [inline tests/features/typing_pep604_test.py:318]: `list[int, str]` isn't a valid type -- was this test meaning to show something else?

File: pyupgrade/_token_helpers.py [hunk 2]
```diff
@@ -166,6 +182,37 @@ def find_block_start(tokens: list[Token], i: int) -> int:
     return i
 
 
+def find_duplicated_types(
```
Comment (asottile) [inline pyupgrade/_token_helpers.py:185]: if this is only used in one file it shouldn't live in the general shared file (same for the other function)

**Review by asottile — COMMENTED** [review]
just a quick first pass -- will look more closely later

File: pyupgrade/_token_helpers.py (hunk 2 above)
Comment (walter9388) [reply pyupgrade/_token_helpers.py]: Yeah fair. I've moved them both to the `typing_604.py` file.

File: tests/features/typing_pep604_test.py (hunk 1 above)
Comment (walter9388) [reply tests/features/typing_pep604_test.py]: That was dumb of me. I had it marked as `xfail` so I didn't pick up on using `,` over `|`. I've updated that example now and I'll update the PR message as well.

File: pyupgrade/_plugins/typing_pep604.py [hunk 3]
```diff
@@ -14,22 +14,36 @@
...
+from pyupgrade._token_helpers import _OPENING
 from pyupgrade._token_helpers import find_op
 from pyupgrade._token_helpers import is_close
 from pyupgrade._token_helpers import is_open
 
 
 def _fix_optional(i: int, tokens: list[Token]) -> None:
     j = find_op(tokens, i, '[')
-    k = find_closing_bracket(tokens, j)
+    k, contains_none = _find_closing_bracket_and_if_contains_none(tokens, j)
```
Comment (walter9388) [inline pyupgrade/_plugins/typing_pep604.py:25]: Modified the general `find_closing_bracket` function to also check for whether the optional block already contains `None` in the same pass.

File: pyupgrade/_plugins/typing_pep604.py [hunk 4]
```diff
@@ -14,22 +14,36 @@
...
     if tokens[j].line == tokens[k].line:
-        tokens[k] = Token('CODE', ' | None')
+        if contains_none:
+            del tokens[k]
+        else:
+            tokens[k:k + 1] = [
+                Token('UNIMPORTANT_WS', ' '),
+                Token('CODE', '| '),
+                Token('CODE', 'None'),
+            ]
```
Comment (walter9388) [inline pyupgrade/_plugins/typing_pep604.py:34]: The reason for changing the single token containing ` | None` to explicit whitespace, `|` and `None` is for the deduplication and whitespace removal functions used in `_fix_union`. This also applies to the multiline version a few lines below.

File: pyupgrade/_plugins/typing_pep604.py [hunk 5]
```diff
@@ -110,11 +134,80 @@ def _fix_union(
 
         for comma in comma_positions:
             tokens[comma] = Token('CODE', ' |')
+        to_delete += _remove_consecutive_unimportant_ws(
+            tokens, [x for x in range(j, k) if x not in to_delete],
+        )
```
Comment (walter9388) [inline pyupgrade/_plugins/typing_pep604.py:139]: Not convinced this is the best approach to remove whitespace, but not sure about what to do in situations where lines are completely deleted other than comments. I have written a niche test for this situation in test `id='duplicated types in multi-line nested unions or optionals'`.

**Review by walter9388 — COMMENTED** [review]
Just highlighting a few things to be aware of.

File: pyupgrade/_plugins/typing_pep604.py (hunk 4 above)
Comment (asottile) [reply pyupgrade/_plugins/typing_pep604.py:34]: it might be helpful to do all the tokens then -- `|` would be `'OP'` and `None` would be `'NAME'`

File: pyupgrade/_plugins/typing_pep604.py [hunk 6]
```diff
@@ -110,11 +134,80 @@ def _fix_union(
...
+        i = d1c
+    return to_delete
+
+
+def _remove_consecutive_unimportant_ws(
+    tokens: list[Token], idxs: list[int],
+) -> list[int]:
+    to_delete = []
+    prev_name = ''
+    for kk in idxs:
```
Comment (asottile) [inline pyupgrade/_plugins/typing_pep604.py:201]: why kk?
